cmd/email-exporter/main.go

```diff
@@ -0,0 +1,103 @@
+package notmain
+
+import (
+	"context"
+	"flag"
+	"os"
+
+	"github.com/letsencrypt/boulder/cmd"
+	"github.com/letsencrypt/boulder/email"
+	emailpb "github.com/letsencrypt/boulder/email/proto"
+	bgrpc "github.com/letsencrypt/boulder/grpc"
+)
+
+// Config holds the configuration for the email-exporter service.
+type Config struct {
+	EmailExporter struct {
+		cmd.ServiceConfig
+
+		// PardotBusinessUnit is the Pardot business unit to use.
+		PardotBusinessUnit string `validate:"required"`
+
+		// ClientId is the OAuth API client ID provided by Salesforce.
+		ClientId cmd.PasswordConfig
+
+		// ClientSecret is the OAuth API client secret provided by Salesforce.
+		ClientSecret cmd.PasswordConfig
+
+		// SalesforceBaseURL is the base URL for the Salesforce API. (e.g.,
+		// "https://login.salesforce.com")
+		SalesforceBaseURL string `validate:"required"`
+
+		// PardotBaseURL is the base URL for the Pardot API. (e.g.,
+		// "https://pi.pardot.com")
+		PardotBaseURL string `validate:"required"`
+	}
+	Syslog        cmd.SyslogConfig
+	OpenTelemetry cmd.OpenTelemetryConfig
+}
+
+func main() {
+	configFile := flag.String("config", "", "Path to configuration file")
+	grpcAddr := flag.String("addr", "", "gRPC listen address override")
+	debugAddr := flag.String("debug-addr", "", "Debug server address override")
+	flag.Parse()
+
+	if *configFile == "" {
+		flag.Usage()
+		os.Exit(1)
+	}
+
+	var c Config
+	err := cmd.ReadConfigFile(*configFile, &c)
+	cmd.FailOnError(err, "Reading JSON config file into config structure")
+
+	if *grpcAddr != "" {
+		c.EmailExporter.ServiceConfig.GRPC.Address = *grpcAddr
+	}
+	if *debugAddr != "" {
+		c.EmailExporter.ServiceConfig.DebugAddr = *debugAddr
+	}
+
+	scope, logger, oTelShutdown := cmd.StatsAndLogging(c.Syslog, c.OpenTelemetry, c.EmailExporter.ServiceConfig.DebugAddr)
+	defer oTelShutdown(context.Background())
+
+	logger.Info(cmd.VersionString())
+
+	clk := cmd.Clock()
+	clientId, err := c.EmailExporter.ClientId.Pass()
+	cmd.FailOnError(err, "Loading client ID")
+	clientSecret, err := c.EmailExporter.ClientSecret.Pass()
+	cmd.FailOnError(err, "Loading client secret")
+
+	pardotClient, err := email.NewPardotClient(
+		clk,
+		c.EmailExporter.PardotBusinessUnit,
+		clientId,
+		clientSecret,
+		c.EmailExporter.SalesforceBaseURL,
+		c.EmailExporter.PardotBaseURL,
+	)
+	cmd.FailOnError(err, "Creating Pardot client")
+	exporterServer := email.NewExporterImpl(pardotClient, scope, logger)
+
+	daemonCtx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// Begin asynchronous processing of the email queue.
+	go exporterServer.Start(daemonCtx)
```

Comment:
By virtue of running this in a goroutine, all of the draining logic gets bypassed: the context will get cancelled, but the `main` goroutine will exit immediately afterwards, killing the whole process before the background goroutine has a chance to drain. Either:
a) This needs to be a synchronous call at the very bottom of `main`, so that the whole process blocks on it draining; or
b) This needs a way to signal back to the main goroutine when it is done. This could follow the example of the RA, which has an explicit [drain](https://github.com/letsencrypt/boulder/blob/27cbd1c94c6c676b64665226502c2f20cd1c4ba3/ra/ra.go#L2603) method [which `main` defers](https://github.com/letsencrypt/boulder/blob/27cbd1c94c6c676b64665226502c2f20cd1c4ba3/cmd/boulder-ra/main.go#L344).